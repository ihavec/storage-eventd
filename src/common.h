#ifndef _COMMON_H_
#define _COMMON_H_
#include <stdlib.h>
#include <stdbool.h>
#include <syslog.h>
#include <libudev.h>
#include <libconfig.h>

#include "list.h"

#ifndef __noreturn
#define __noreturn      __attribute__((noreturn))
#endif

struct global_state {
	char hostname[128];
	const char *program_name;
	const char *config_file;
	bool foreground;
	bool dry_run;
	bool debug;
	int verbose;
	volatile bool exiting;
	uid_t unpriv_uid;
	gid_t unpriv_gid;
};
extern struct global_state global_state;

void error(const char *fmt, ...) __noreturn;
void assert_failure(const char *file, int line, const char *cond) __noreturn;
void log_msg(int level, const char *fmt, ...);

#define log_err(fmt, args...)		log_msg(LOG_ERR, fmt, ##args)
#define log_warn(fmt, args...)		log_msg(LOG_WARNING, fmt, ##args)
#define log_info(fmt, args...)		log_msg(LOG_INFO, fmt, ##args)
#define log_debug(fmt, args...)		log_msg(LOG_DEBUG, fmt, ##args)

#define ASSERT(condition)					\
do {								\
	if (!(condition))					\
		assert_failure(__FILE__, __LINE__, #condition);	\
} while(0)

static inline void *ERR_PTR(long error)
{
	return (void *) error;
}

static inline long PTR_ERR(const void *ptr)
{
	return (long) ptr;
}

static inline void *zalloc(size_t size)
{
	return calloc(1, size);
}

#define ARRAY_SIZE(a)	(sizeof(a)/sizeof(a[0]))

#endif /* COMMON_H_ */
