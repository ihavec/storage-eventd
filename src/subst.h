#ifndef _SUBST_H_
#define _SUBST_H_
#include <stdbool.h>
#include <sys/types.h>
struct subst {
	bool literal;
	const char *value;
	size_t len;
};

struct subst_vec {
	int size;
	int count;
	struct subst vec[0];
};

typedef const char *(*subst_lookup_fn)(const char *key, void *data,
				      bool *needs_free);

int subst_tokenize(const char *str, struct subst_vec **vecp);
void subst_release(struct subst_vec *vec);
const char *subst_replace(const struct subst_vec *vec, subst_lookup_fn lookup,
			  void *data);
#endif /*_SUBST_H_ */
