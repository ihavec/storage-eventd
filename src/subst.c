#define _GNU_SOURCE
#include <string.h>
#include <errno.h>
#include <stdlib.h>
#include <glib.h>
#include <string.h>
#include <stdio.h>

#include "common.h"
#include "subst.h"

static struct subst_vec *
resize_subv(struct subst_vec *vec, int size)
{
	return realloc(vec, sizeof(*vec) + size * sizeof(vec->vec[0]));
}

/*
 * Return the next token structure.  Resize the vector 10 slots at a time.
 * Optimize for the common case and use 2 slots.
 */
static struct subst *
get_sub(struct subst_vec **vecp)
{
	struct subst_vec *vec = *vecp;
	if (!vec) {
		vec = resize_subv(NULL, 2);
		vec->size = 2;
		vec->count = 0;
	} else if (vec->size == vec->count)
		vec = resize_subv(vec, vec->size + 10);

	*vecp = vec;
	return &vec->vec[vec->count++];
}

/*
 * Break up a string into literal and substitition tokens.  The string
 * is not modified or copied and will need to be available for the lifetime
 * of the subst_vec.
 */
int
subst_tokenize(const char *str, struct subst_vec **vecp)
{
	struct subst_vec *subv = NULL;
	struct subst *sub = NULL;
	const char *cur = str;
	int count = 0;
	int deep = 0;

	while (*cur) {
		if (!sub) {
			sub = get_sub(&subv);
			sub->value = cur;
		}
		if (cur[0] == '\\' && cur[1] == '$') {
			cur += 2;
			continue;
		} else if (cur[0] == '$' && cur[1] == '{') {
			if (cur == sub->value) {
				sub->value += 2;
				deep = 1;
			} else if (!deep++) {
				sub->len = cur - sub->value;
				sub->literal = true;
				sub = NULL;
			} else {
				log_err("syntax error at position %u: nested variables are not allowed.",
					cur - str);
				goto fail;
			}
			cur += 2;
			continue;
		} else if (*cur == '}') {
			/*
			 * We actually could handle nested variables but
			 * we don't currently have a way to order them
			 * properly.
			 */
			if (deep > 0) {
				sub->len = cur - sub->value;
				sub->literal = false;
				sub = NULL;
				--deep;
			}
		} else if (*cur == ' ' && deep) {
			log_err("syntax error at position %u: spaces not allowed in variable names.",
				cur - str);
			goto fail;
		}
		cur++;
	}
	if (sub) {
		if (deep) {
			log_err("parse error at end of input: missing closing `}' for variable substitution.");
			goto fail;
		}
		sub->literal = true;
		sub->len = cur - sub->value;
	}

	*vecp = resize_subv(subv, subv->count);

	if (global_state.verbose > 2)
		for (count = 0; count < subv->count; count++) {
			log_debug("subv->vec[%d] = \"%s%.*s\" [%u]", count,
				  subv->vec[count].literal ? "" : "$",
				  subv->vec[count].len,
				  subv->vec[count].value,
				  subv->vec[count].len);
	}

	return 0;

fail:
	subst_release(subv);
	*vecp = NULL;
	return -EINVAL;
}

/*
 * Reconstitute the subst vector as a single string.  The caller is responsible
 * for freeing it.  If a variable can't be found using the lookup routine
 * provided, the string ${VARIABLE} using the variable name will be
 * used instead.
 *
 * All of the allocations except for the final product are done on the stack.
 */
const char *
subst_replace(const struct subst_vec *vec, subst_lookup_fn lookup, void *data)
{
	int i;
	char *ret, *str;
	int len = 0;
	const char *props[vec->count];
	bool needs_free[vec->count];
	int lengths[vec->count];

	/* A string literal */
	if (vec->count == 1 && vec->vec[0].literal) 
		return strndup(vec->vec[0].value, vec->vec[0].len);

	memset(props, 0, sizeof(props));


	for (i = 0; i < vec->count; i++) {
		const char *key;
		if (vec->vec[i].literal) {
			len += vec->vec[i].len;
			continue;
		}
		key = strndupa(vec->vec[i].value, vec->vec[i].len);
		props[i] = lookup(key, data, &needs_free[i]);

		/*
		 * If there's a lookup failure, we output the variable name
		 * as if the specification were a string literal.
		 */
		if (props[i] == NULL) {
			char *buf = alloca(vec->vec[i].len + 4);
			snprintf(buf, vec->vec[i].len + 4, "${%s}", key);
			props[i] = buf;
		}
		lengths[i] = strlen(props[i]);
		len += lengths[i];
	}

	/*
	 * Reassemble the vector as a single string, using the lookups we
	 * just performed.
	 */

	str = ret = malloc(len + 1);
	if (!str)
		return NULL;

	memset(str, 0, len + 1);
	for (i = 0; i < vec->count; i++) {
		if (vec->vec[i].literal) {
			g_assert(vec->vec[i].len <= len);
			strncat(str, vec->vec[i].value, vec->vec[i].len);
			len -= vec->vec[i].len;
		} else {
			g_assert(lengths[i] <= len);
			strncat(str, props[i], lengths[i]);
			len -= lengths[i];
		}
		if (needs_free[i])
			free((char *)props[i]);
	}
	g_assert(len == 0);

	return ret;
}

void
subst_release(struct subst_vec *vec)
{
	free(vec);
}
