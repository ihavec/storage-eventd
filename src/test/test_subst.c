#include <stdio.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/fcntl.h>
#include <unistd.h>

#include "common.h"
#include "subst.h"

struct test_case {
	const char *string;
	int expected_ret;
} test_cases[] = {
	{ "${VALUE}", 0 },
	{ "${VA${FOO}LUE}", 0 },
	{ "${VA${FOO}LUE", -EINVAL },
	{ "Test then open variable ${VA${FOO}LUE", -EINVAL },
	{ "${VA${FOO}LUE and more text", -EINVAL },
	{ "Test then open variable ${VA${FOO}LUE and more text", -EINVAL },
	{ "This is a test ${VALUE}", 0 },
	{ "This is a test ${VALUE} with text after", 0 },
	{ "This is a test ${VALUE} with text after and another ${VAR}", 0 },
	{ "This is a test ${VALUE} with text after and another ${VAR} text", 0 },
	{ "This is a test $VALUE} with text after and another ${VAR} text", 0 },
	{ "This is an ${invalid variable}", -EINVAL },
};

static int
test_one(const char *string, int expected_ret)
{
	int ret;
	struct subst_vec *vec = NULL;

	printf("C: \"%s\"\n", string);
	ret = subst_tokenize(string, &vec);

	if (ret == 0) {
		int count;
		for (count = 0; count < vec->count; count++) {
			struct subst *sub = &vec->vec[count];
			printf("%u: %c/\"%.*s\"\n", count,
				sub->literal ? 'L' : 'V',
				sub->len, sub->value);
		}
		subst_release(vec);
	}

	if (ret != expected_ret) {
		printf("test `%s' failed.\n", string);
		printf("---\n");
		return -1;
	}
	printf("---\n");
	return 0;
}

#define DEFAULT_TEMPLATE_FILE BASEDIR "/src/email-template.txt"
int
main(void)
{
	int fd;
	int i;
	char *p;
	struct stat st;
	for (i = 0; i < ARRAY_SIZE(test_cases); i++)
		test_one(test_cases[i].string, test_cases[i].expected_ret);

	fd = open(DEFAULT_TEMPLATE_FILE, O_RDONLY);
	fstat(fd, &st);
	p = malloc(st.st_size + 1);
	read(fd, p, st.st_size);
	p[st.st_size] = 0;
	close(fd);
	test_one(p, 0);

	return 0;
}
