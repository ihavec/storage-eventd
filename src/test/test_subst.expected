parse error at end of input: missing closing `}' for variable substitution.
parse error at end of input: missing closing `}' for variable substitution.
parse error at end of input: missing closing `}' for variable substitution.
parse error at end of input: missing closing `}' for variable substitution.
parse error at end of input: missing closing `}' for variable substitution.
C: "${VALUE}"
0: V/"VALUE"
---
C: "${VA${FOO}LUE}"
0: V/"VA${FOO}LUE"
---
C: "${VA${FOO}LUE"
---
C: "Test then open variable ${VA${FOO}LUE"
---
C: "${VA${FOO}LUE and more text"
---
C: "Test then open variable ${VA${FOO}LUE and more text"
---
C: "This is a test ${VALUE}"
0: L/"This is a test "
1: V/"VALUE"
---
C: "This is a test ${VALUE} with text after"
0: L/"This is a test "
1: V/"VALUE"
2: L/" with text after"
---
C: "This is a test ${VALUE} with text after and another ${VAR}"
0: L/"This is a test "
1: V/"VALUE"
2: L/" with text after and another "
3: V/"VAR"
---
C: "This is a test ${VALUE} with text after and another ${VAR} text"
0: L/"This is a test "
1: V/"VALUE"
2: L/" with text after and another "
3: V/"VAR"
4: L/" text"
---
C: "This is a test $VALUE} with text after and another ${VAR} text"
test `This is a test $VALUE} with text after and another ${VAR} text' failed.
---
